Core/src/com/serotonin/m2m2/rt/script/CompiledMangoJavaScript.java

```diff
@@ -0,0 +1,232 @@
+            }else {
+                toCompile = script;
+            }
+            
+            this.compiledScript = ((Compilable)engine).compile(toCompile);
+        }catch(ScriptException e) {
+            throw ScriptError.create(e);
+        }
+    }
+    
+    /**
+     * Clear the engine scope and initialize it
+     * 
+     * @throws ScriptException
+     */
+    public void initialize() throws ScriptException {
+        Bindings engineScope = compiledScript.getEngine().getBindings(ScriptContext.ENGINE_SCOPE);
+        if(setter == null)
+            setter = service.createValidationSetter(result, permissions);
+        
+        engineScope.put(MangoJavaScriptService.UNCHANGED_KEY, MangoJavaScriptService.UNCHANGED);
+        List<String> points = new ArrayList<String>();
+        engineScope.put(MangoJavaScriptService.POINTS_CONTEXT_KEY, points);
+        engineScope.put(MangoJavaScriptService.TIMESTAMP_CONTEXT_KEY, null);
+        
+        if(permissions != null) {
+            utilities.clear();
+            for(MangoJavascriptContextObjectDefinition def : ModuleRegistry.getMangoJavascriptContextObjectDefinitions()) {
+                ScriptUtility util = testRun ? def.initializeTestContextObject(permissions) : def.initializeContextObject(permissions);
+                util.setScriptLog(log);
+                util.setResult(result);
+                util.takeContext(compiledScript.getEngine(), engineScope, setter, importExclusions, testRun);
+                engineScope.put(util.getContextKey(), util);
+                utilities.add(util);
+            }
+            //Initialize additional utilities
+            for(ScriptUtility util : additionalUtilities) {
+                util.setScriptLog(log);
+                util.setResult(result);
+                util.takeContext(compiledScript.getEngine(), engineScope, setter, importExclusions, testRun);
+                engineScope.put(util.getContextKey(), util);
+            }
+        }
+        
+        Set<Entry<String,Object>> entries = additionalContext.entrySet();
+        for(Entry<String,Object> entry: entries)
+            engineScope.put(entry.getKey(), entry.getValue());
+                
+        // Put the context variables into the engine with engine scope.
+        if(context != null) {
+            for (String varName : context.keySet()) {
+                IDataPointValueSource point = context.get(varName);
+                engineScope.put(varName, service.wrapPoint(compiledScript.getEngine(), point, setter));
+                if (points != null)
+                    points.add(varName); 
+            }
+            
+            engineScope.put(MangoJavaScriptService.POINTS_MAP_KEY, context);
+        }
+        
+        //Set the print writer and log
+        compiledScript.getEngine().getContext().setWriter(log.getStdOutWriter());
```

Comment:
What was the ambition here? I think this will always be a null writer as constructed, but I wonder if this would lead to outputting print() or errors to the log, which may or may not be desired.

Reply:
The general idea is that when testing a script you can save the stdout info.  If you follow the `ScriptLog` constructor from MangoJavaScriptService.testScript() you will see that it creates a non-rolling log and then the PrintWriter is set to stdout so we can capture the output during a script validation.